[MRG] Add elastic net penalty to LogisticRegression

NicolasHug opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#pull-request-checklist
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->

Closes #8288

#### What does this implement/fix? Explain your changes.

This PR allows the use of `penalty='elastic-net'` for the `LogisticRegression` class. Elastic net is already available in the saga solver, it's just not exposed yet.

#### Any other comments?

~~I set the `l1_ratio` parameter to 0.5 by default, because that's the default in `ElasticNet`. Should I set it to `None`? This would allow to raise an error if it's set to something else while `penalty != 'elastic-net'`.~~ 
EDIT: I set it to None and issue warning if penalty != elastic-net

#### TODO

- [X] expose `penalty='elastic-net'` in `LogisticRegression`
- [X] expose `penalty='elastic-net'` in `LogisticRegressionCV`
- [X] Update doc
- [X] write tests
- [X] Update examples

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

NicolasHug committed:
First draft on elasticnet penaly for LogisticRegression

NicolasHug committed:
Some basic tests

NicolasHug committed:
Doc update

amueller commented:
So you want to set it to None but have the behavior be 0.5? I don't have a strong feeling about it.

Otherwise looks like it's going in the right direction :)

NicolasHug committed:
First draft for LogisticRegressionCV.

It seems to be working for binary classification and for multiclass when
multi_class='ovr'. I'm having a hard time figuring out the intricacies
of multi_class='multinomial'.

NicolasHug commented:
I just committed a first shot at LogisticRegressionCV.

I looks OK for binary classification and multiclass when multiclass='ovr', but does not work for multiclass='multinomial'.

I'm pretty sure that the problem comes from [this line](https://github.com/NicolasHug/scikit-learn/blob/5cb41c460b2601f6c2035b998b4baaeda8bb2020/sklearn/linear_model/logistic.py#L1794), which was fine when there was only one parameter to optimize (C), but now it messes up with the format of the scores array. So the computation of the best_index (index of the best parameters), which depends on the score array, is incorrect.

I tried various ways of dealing with this but every single change I tried would break other parts of the code.

I'll focus on the doc and the examples for now, but any help / insight would be helpful :)

NicolasHug committed:
Changed default to None for l1_ratio.

added warning message is user sets l1_ratio while penalty is not
elastic-net

NicolasHug committed:
Some more doc

NicolasHug committed:
Updated example to plot elastic net sparsity

NicolasHug committed:
Fixed flake8

NicolasHug committed:
Fixed test by not modifying attribute in fit

NicolasHug committed:
Fixed doc issues

NicolasHug committed:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn into elastic_net_linear_classifier

NicolasHug commented:
One of the doctests in 3.6 if failing simply because of a different formatting for floats in numpy arrays... Is there any clean way to ignore it? I guess changing the expected output string would break the other versions tests...

jnothman commented:
we only do those doctests with the latest numpy, so you should conform

NicolasHug committed:
WIP

NicolasHug committed:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn into elastic_net_linear_classifier

NicolasHug committed:
Partially fixed logistic_reg_CV for multinomial.

Also added some comments that are hopefully clear.
Still need to fix refit=False

NicolasHug committed:
Fixed doc issue

NicolasHug committed:
WIP

NicolasHug committed:
Fixed test for refit=False in LogisticRegressionCV

NicolasHug committed:
Fixed Python 2 numpy version issue

NicolasHug committed:
minor doc updates

NicolasHug committed:
Weird doc error...

NicolasHug commented:
Marking this as MRG for a first review, looks like LogisticRegressionCV is finally working as expected \o/

NicolasHug committed:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn into elastic_net_linear_classifier

jnothman commented:
please ping after the upcoming release

NicolasHug commented on sklearn/linear_model/logistic.py:
```diff
@@ -1706,69 +1784,88 @@ def fit(self, X, y, sample_weight=None):
+                # For multinomial, all scores are the same accross classes
+                scores = scores[0]
+                # coefs_paths will keep its original shape because
+                # logistic_regression_path excepts it this way
```
expects

amueller commented on sklearn/linear_model/tests/test_logistic.py:
```diff
@@ -224,6 +226,15 @@ def test_check_solver_option(LR):
         lr = LR(solver=solver, dual=True)
         assert_raise_message(ValueError, msg, lr.fit, X, y)
 
+    # only saga supports elastic-net. We only test for linlinear because the
```
liblinear

amueller commented on sklearn/linear_model/tests/test_logistic.py:
```diff
@@ -1309,6 +1320,202 @@ def test_warm_start_converge_LR():
+
+
+def test_ElasticNetCV_no_refit():
+    print()
```
?

amueller commented:
can we also test that the solution is sensible in some way, not only different from the l1 and l2 ones? Maybe grid-searching l1_ratio should be at least as good as using just l1 or l2?
And/Or we could check that the elasticnet objective is actually lower when using the elasticnet regularizer than when using l1 or l2?

NicolasHug committed:
Added test to ensure that elastic net is at least as good as L1 or L2
once l1_ratio has been optimized with grid search

Also addressed minor reviews

NicolasHug commented:
Thanks! I just added a test for the first suggestion. Unfortunately elastic net does not seem to be strictly better, hence the `>=` instead of `>`. I tried to play a bit with the parameters of `make_classification` but got the same results.

> we could check that the elasticnet objective is actually lower when using the elasticnet regularizer than when using l1 or l2

I am not sure why this would be true, could you please elaborate?

NicolasHug committed:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn into elastic_net_linear_classifier

NicolasHug committed:
Fixed test

NicolasHug commented:
pinging @amueller @jnothman :)

amueller commented on sklearn/linear_model/tests/test_logistic.py:
```diff
@@ -78,6 +80,7 @@ def test_predict_2_classes():
                                          random_state=0), X_sp, Y1)
 
 
+@pytest.mark.filterwarnings('ignore: Default solver will be changed')  # 0.22
```
Why are these required here? That's a bit confusing to me.

amueller replied:
I mean why has this become necessary with your change and not before?

NicolasHug replied:
It's quite mysterious for me too, the test passes without the `filterwarning` on my machine but not on travis

amueller commented on sklearn/linear_model/tests/test_logistic.py:
```diff
@@ -1379,6 +1392,220 @@ def test_warm_start_converge_LR():
+
+
+def test_elastic_net_CV_grid_search():
+    # make sure ElasticNetCV gives same best params as GridSearchCV
```
LogisticRegressionCV with elastic net? There's also ElasticNetCV which is regression

amueller commented on sklearn/linear_model/logistic.py:
```diff
@@ -1785,69 +1862,88 @@ def fit(self, X, y, sample_weight=None):
                 scores = self.scores_[cls]
                 coefs_paths = self.coefs_paths_[cls]
+            else:
+                # For multinomial, all scores are the same accross classes
```
across

amueller commented on sklearn/linear_model/tests/test_logistic.py:
```diff
@@ -1379,6 +1392,220 @@ def test_warm_start_converge_LR():
+
+
+def test_elastic_net_CV_multiclass_multinomial():
+    # make sure ElasticNetCV gives same best params as GridSearchCV with
```
same comment about ElasticNetCV.

amueller reviewed (commented):
Looks good. Can we maybe actually have a functional test in some way? Would it make sense to compare against SGDClassifier with elasticnet penalty?

jnothman commented on sklearn/linear_model/logistic.py:
```diff
@@ -781,9 +793,13 @@ def logistic_regression_path(X, y, pos_class=None, Cs=10, fit_intercept=True,
+            elif penalty == 'l2':
                 alpha = 1. / C
                 beta = 0.
+            else:  # elastic-net penalty (alpha is 2-norm, beta is 1-norm)
```
I think you can make this comment before the `if`

jnothman commented on sklearn/linear_model/logistic.py:
```diff
@@ -1607,6 +1658,9 @@ class LogisticRegressionCV(LogisticRegression, BaseEstimator,
+    l1_ratios_ : array
+        Array of l1_ratios used for cross-validation.
+
     coefs_paths_ : array, shape ``(n_folds, len(Cs_), n_features)`` or \
```
Shouldn't there be one for each `l1_ratio`? Surely l1_ratios should become another dimension. See also shape of `ElasticNetCV.mse_paths_` (which is quite different... and I'm not really sure why we're storing all the coefs; it seems expensive)

jnothman replied:
Similarly, the shape mentioned in `scores_` below shoulld be updated.

NicolasHug replied:
I don't understand what you mean. `l1_ratios_` is just the 'processed' version of the `l1_ratios` parameter, so it's just an array of floats. Just like the `Cs_` attribute right above.

jnothman replied:
But there should be a distinct vector of coefficients for each C and l1_ratio so the shapes documented here should be updated

NicolasHug replied:
Yes absolutely, sorry I totally misread your comment.
The new shape is ``(n_folds, len(Cs_) x len(l1_ratios), n_features)``, I assume this is more correct as adding a dimension could break users code?

jnothman reviewed (commented):
Please make sure those Attributes shapes are updated and checked by tests.

amueller commented:
>>we could check that the elasticnet objective is actually lower when using the 
 elasticnet regularizer than when using l1 or l2

> I am not sure why this would be true, could you please elaborate?

I now thing comparing to SGD with elasticnet might be a better idea, but the idea I described here was that the coefficients that result from optimizing the elasticnet objective should result in a lower value of the elasticnet objective than the result of optimizing the l2 objective evaluated using the elasticnet objective.

amueller commented:
merge conflicts :-/

NicolasHug committed:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn into elastic_net_linear_classifier

NicolasHug committed:
addressed comments

NicolasHug commented:
I just pushed new changes addressing the comments, however I tried checking the best values of l1_ratio and C from LogisticRegression and SGD were the same, but my test doesn't pass :/

```python

def test_LogisticRegression_vs_SGDClassifier_elastic_net():
    X, y = make_classification(n_samples=1000, n_classes=2, random_state=0)
    cv = StratifiedKFold(5, random_state=0)

    l1_ratios = np.linspace(0, 1, 5)
    Cs = np.logspace(-4, 4, 5)
    alphas = 1 / Cs

    lr = LogisticRegression(penalty='elastic-net', solver='saga',
                            random_state=0)
    param_grid = {'C': Cs, 'l1_ratio': l1_ratios}
    lrgs = GridSearchCV(lr, param_grid, cv=cv)
    lrgs.fit(X, y)

    param_grid = {'alpha': alphas, 'l1_ratio': l1_ratios}
    from sklearn.linear_model import SGDClassifier
    sgd = SGDClassifier(penalty='elasticnet', loss='log',
                        random_state=0)
    gs = GridSearchCV(sgd, param_grid, cv=cv)
    gs.fit(X, y)


    print(gs.best_params_['l1_ratio'], lrgs.best_params_['l1_ratio'])
    print(gs.best_params_['alpha'], 1 / lrgs.best_params_['C'])
    assert gs.best_params_['l1_ratio'] == lrgs.best_params_['l1_ratio']
    assert gs.best_params_['alpha'] == 1 / lrgs.best_params_['C']

```

NicolasHug committed:
Added back ignore warning on tests

amueller commented:
My idea was not to do the grid-search but just compare coefficients. You probably need to make sure the data is scaled for SGD to work and make sure you let it run long enough.

NicolasHug commented:
>I now thing comparing to SGD with elasticnet might be a better idea, but the idea I described here was that the coefficients that result from optimizing the elasticnet objective should result in a lower value of the elasticnet objective than the result of optimizing the l2 objective evaluated using the elasticnet objective.

So if I get it right:

- run LogisticRegression with elasticnet penalty, get coefficients coeffs_1 and compute elastic net objective -> get value_1
- run SGD with L2 penalty, get coefficients coeffs_2 and compute elastic net objective -> get value_2

We should have value_2 > value_1.

That makes sense but I don't understand why we shoud use SGD instead of LogisticRegression?

amueller commented:
Sorry I was suggesting two different tests. Either test that SGD And LogisticRegression get the same coefficients, or test the objective of elasticnet vs l2.

NicolasHug committed:
Added a functional test

NicolasHug commented:
I just added a tests that checks the elastic-net objective values. To be fair I had to tweak the values of C, l1_ratios and the number of relevant / repeated / random features for it to work, so I'm not sure how relevant it is.

As for comparing coefficients with SGDClassifier, I tried. The problem is that SGDClassifier results in much sparser coefficients than LogisticRegression. This is because SGDClassifier uses a much more aggressive regularization scheme than the vanilla l1 or elastic net regularization. As soon as a coefficient crosses 0, it will be bound to 0 forever (see [here](https://github.com/scikit-learn/scikit-learn/blob/a7e17117bb15eb3f51ebccc1bd53e42fcb4e6cd8/sklearn/linear_model/sgd_fast.pyx#L743) and original paper in docstring)


Also, I should probably change the parameter name from 'elastic-net' to 'elasticnet' right? SGDClassifier uses 'elasticnet'.

NicolasHug committed:
Scale data in test... Now failing

NicolasHug committed:
elastic-net --> elasticnet

amueller commented on sklearn/linear_model/tests/test_logistic.py:
```diff
@@ -1379,6 +1394,247 @@ def test_warm_start_converge_LR():
+    lr_l2.fit(X, y)
+
+    def enet_objective(lr):
+        obj = C * log_loss(y, lr.predict(X))
```
needs to be ``predict_proba``

NicolasHug replied:
oops thanks!
Still failing though :/ , same for SGDClassifier

agramfort replied:
it should be lr.decision_function

amueller commented:
I conclude there's something fishy.

If you look at the output of 
```
from sklearn.datasets import make_classification
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import scale
from sklearn.metrics import log_loss

# train a logistic regression with l2 (a) and elasticnet (b) penalties,
# and compute the elasticnet objective. That of a sould be greater than
# that of b.
C = 0.01
l1_ratio = 1
X, y = make_classification(n_samples=10000, n_classes=2, n_features=20,
                         n_informative=10, n_redundant=0,
                         n_repeated=0, random_state=0)
X = scale(X)

lr_enet = LogisticRegression(penalty='elasticnet', solver='saga',
                           random_state=0, C=C, l1_ratio=l1_ratio,
                           verbose=10, tol=1e-8, max_iter=100)
lr_l2 = LogisticRegression(penalty='l2', solver='saga', random_state=0,
                         C=C, verbose=10, tol=1e-8, max_iter=100)
lr_enet.fit(X, y)
lr_l2.fit(X, y)

def enet_objective(lr):
  obj = C * log_loss(y, lr.predict_proba(X))
  obj += l1_ratio * np.linalg.norm(lr.coef_, 1)
  obj += ((1 - l1_ratio) / 2) * np.linalg.norm(lr.coef_, 2)
  return obj
```
then you can see that lr_enet converges for ``l1_ratio==0`` and ``l1_ratio==1`` but doesn't converge (to 1e-7) for any other l1 ratios.

amueller commented:
(also the test passes as expected for l1_ratio=1).

amueller commented:
ping @arthurmensch @GaelVaroquaux @agramfort Are you sure the implementation is still correct if both alpha and beta are non-zero?

amueller commented:
Ok so ``penalty='l2'`` has this as objective function (it's equivalent to what l-bfgs uses):
```python
def l2_objective(lr):
    obj = log_loss(y, lr.predict_proba(X)) * X.shape[0]
    obj += (1 / (2 * C)) * np.linalg.norm(lr.coef_, 2) ** 2
    return obj
```
So I thought ``penalty='l1'`` has this as objective function:
```python
def l1_objective(lr):
    obj = log_loss(y, lr.predict_proba(X)) * X.shape[0]
    obj += (1 / C) * np.linalg.norm(lr.coef_, 1)
    return obj
```
but that doesn't match what liblinear or saga are doing...

amueller commented:
So yeah, it would also be great if someone could provide the objective function for l1 logistic regression because what I wrote doesn't match what liblinear is doing or what saga is doing.

NicolasHug commented:
The closest I got to is 

```
def l1_objective(lr):
    obj = C * log_loss(y, lr.predict_proba(X)) * X.shape[0]
    obj +=  np.linalg.norm(lr.coef_, 1)
    return obj
```

but it's still off. The relative error is still big for small values of `C`.

GaelVaroquaux commented:
> but it's still off. The relative error is still big for small values of C.

Could it be the intercept that's penalized in liblinear?

NicolasHug commented:
We tried but adding the intercept doesn't have much effect as it's usually very small

NicolasHug committed:
Updated doc for some attributes and checked their shape in tests

jnothman commented:
Pressed the wrong button. Add a dimension. It won't break users' code as they don't set l1_ratios

NicolasHug committed:
Added l1_ratio dimension to coefs_paths and scores attr

amueller commented:
@jnothman do you have any insights on the convergence issues or loss function. I'm a bit stumped.

jnothman commented:
I'm not sure I will have time to attend to this for a while...

amueller commented:
@agramfort your thoughts here would be much appreciated. I still can't really figure out what's happening in the l1 case as what I thought the objective was doesn't match was saga optimizes and doesn't match what liblinear computes.

agramfort commented on sklearn/linear_model/tests/test_logistic.py:
```diff
@@ -1379,6 +1394,272 @@ def test_warm_start_converge_LR():
+    def enet_objective(lr):
+        obj = C * log_loss(y, lr.predict(X))
+        obj += l1_ratio * np.linalg.norm(lr.coef_, 1)
+        obj += ((1 - l1_ratio) / 2) * np.linalg.norm(lr.coef_, 2)
```
np.linalg.norm(lr.coef_, 2) should be np.linalg.norm(lr.coef_, 2)**2

agramfort commented on sklearn/linear_model/tests/test_logistic.py:
```diff
@@ -1379,6 +1394,272 @@ def test_warm_start_converge_LR():
+
+    lr_enet = LogisticRegression(penalty='elasticnet', solver='saga',
+                                 random_state=0, C=C, l1_ratio=l1_ratio)
+    lr_l2 = LogisticRegression(penalty='l2', solver='saga', random_state=0,
```
if you use `fit_intercept=False` test passes but I am not sure of why this test makes sense. What is the rationale for assuming that 

`l1_ratio * np.linalg.norm(lr.coef_, 1) < l1_ratio/2 * np.linalg.norm(lr.coef_, 2)**2``

for any value of coef_

what do I miss?

NicolasHug replied:
We're not checking that                                                        
                                                                               
`l1_ratio * np.linalg.norm(lr.coef_, 1) < l1_ratio/2 * np.linalg.norm(lr.coef_, 2)**2`
                                                                               
we're training a logistic regression with l2 (a) and a logistic regression with elasticnet (b), and check that the elasticnet objective (which includes the predictions) of a is higher than that of b. I tried to illustrate this on a drawing (see details)

<details>

![pice](https://user-images.githubusercontent.com/1190450/47570859-2d20fa80-d905-11e8-8dd8-f454fc7ab7dd.jpg)


</details>

agramfort committed:
improve example + fix test

TomDLT commented:
I tried to find what is actually optimized in `LogisticRegression` with elastic-net.
I found that the following function is correct in the L1 and in the L2 case, but not in the ElasticNet case.
```py
    def objective(lr):
        coef = lr.coef_.ravel()
        obj = log_loss(y, lr.predict_proba(X)) * n_samples
        obj += l1_reg * np.sum(np.abs(coef))
        obj += l2_reg * 0.5 * np.dot(coef, coef)
        return obj
```
Note that I don't use `np.linalg.norm` which is not precise enough.
I also checked this objective with our `SGDClassifier`, and with lightning's `SAGAClassifier`. 
- The objective is correct in the L1 and L2 cases, and all three estimators give similar results, though SGD is much slower to converge, as expected.
- In the elastic-net case, the objective seems correct for `SAGAClassifier` and `SGDclassifier`, which give similar results. However, the objective is not correct for `LogisticRegression`.
- Multiplying the L2 regularization by 2 for `LogisticRegression` (only when `l1-ratio > 0`) somewhat get similar results with the two other estimators, but it does not converge to the exact same point, though it does in the L1 and L2 cases.

I guess our SAGA implementation is not exact for the elastic-net case.

<details>

```py
import numpy as np
import matplotlib.pyplot as plt

from sklearn.datasets import make_classification
from sklearn.linear_model import LogisticRegression
from sklearn.linear_model import SGDClassifier
from sklearn.preprocessing import scale
from sklearn.metrics import log_loss
from sklearn.utils.testing import ignore_warnings

from lightning.classification import SAGAClassifier

###############################################################################
# Dataset
n_samples = 1000
X, y = make_classification(n_samples=n_samples, n_classes=2, n_features=5,
                           n_informative=5, n_redundant=0, n_repeated=0,
                           random_state=0)
X = scale(X)

###############################################################################
# Regularization
C = 0.01
for l1_ratio in [0, 0.1, 0.5, 0.9, 1]:

    l2_reg = (1. / C) * (1 - l1_ratio)
    l1_reg = (1. / C) * l1_ratio
    print("L1: %.2f, L2: %.2f" % (l1_reg, l2_reg))

    # parameters for SGD:
    alpha_sgd = (l2_reg + l1_reg) / n_samples
    l1_ratio_sgd = l1_reg / (l2_reg + l1_reg)

    # parameters for lightning's SAGAClassifier
    alpha_lightning = l2_reg / n_samples
    beta_lightning = l1_reg / n_samples

    # parameters for LogisticRegression
    # XXX: weird, somewhat helps getting a rough equivalence
    if True:
        l2_reg_ = l2_reg * 2 if l1_reg != 0 else l2_reg
    else:
        l2_reg_ = l2_reg

    C_logistic = 1. / (l2_reg_ + l1_reg)
    l1_ratio_logistic = l1_reg / (l2_reg_ + l1_reg)

    def objective(lr):
        coef = lr.coef_.ravel()
        obj = log_loss(y, lr.predict_proba(X)) * n_samples
        obj += l1_reg * np.sum(np.abs(coef))
        obj += l2_reg * 0.5 * np.dot(coef, coef)
        return obj

    ###########################################################################
    # Benchmark: (name, max_iter, estimator)
    methods = [
        ('SGD', 10000, SGDClassifier(
            penalty='elasticnet', loss='log', random_state=0, alpha=alpha_sgd,
            l1_ratio=l1_ratio_sgd, tol=-np.inf, fit_intercept=False), ),
        ('lightning', 100,
         SAGAClassifier(alpha=alpha_lightning, beta=beta_lightning, loss='log',
                        tol=-1, penalty='l1', random_state=0), ),
        ('LogReg', 100, LogisticRegression(
            penalty='elasticnet', solver='saga', random_state=0, C=C_logistic,
            l1_ratio=l1_ratio_logistic, tol=0, fit_intercept=False), ),
    ]

    results = []
    for name, max_max_iter, estimator in methods:
        max_iter_range = np.int_(np.linspace(1, max_max_iter, 10))

        objective_list = []
        for max_iter in max_iter_range:
            print('.', end='', flush=True)
            estimator.max_iter = max_iter
            with ignore_warnings():
                estimator.fit(X, y)
            objective_list.append(objective(estimator))

        coef = estimator.coef_.ravel()
        print("coefs: %s (%s)" % (coef, name))
        results.append(np.array(objective_list))

    results = np.array(results)
    fig = plt.figure()
    plt.semilogy(max_iter_range, results.T - np.min(results) + 1e-15, '.-')
    plt.legend([name for name, _, _ in methods])
    plt.title('l1_ratio = %s' % l1_ratio)
    plt.xlabel('n_iter (SAGA) or n_iter / 100 (SGD)')

    fig.savefig('%.1f.png' % l1_ratio)
    plt.close('all')
```
___
**Without the factor 2:**
```
L1: 0.00, L2: 100.00
..........coefs: [ 0.32072801 -1.14099813  0.24350405 -0.45900174 -0.13345408] (SGD)
..........coefs: [ 0.32072806 -1.14099771  0.24350422 -0.45900164 -0.13345406] (lightning)
..........coefs: [ 0.32072806 -1.14099771  0.24350422 -0.45900164 -0.13345406] (LogReg)
L1: 10.00, L2: 90.00
..........coefs: [ 0.28944688 -1.14915421  0.21572344 -0.44261878 -0.09801055] (SGD)
..........coefs: [ 0.28944694 -1.14915376  0.21572362 -0.44261866 -0.09801052] (lightning)
..........coefs: [ 0.31517452 -1.50523317  0.21895128 -0.59221208 -0.14362203] (LogReg)
L1: 50.00, L2: 50.00
..........coefs: [ 0.14099022 -1.21314473  0.08441194 -0.36694632  0.        ] (SGD)
..........coefs: [ 0.14099785 -1.21312388  0.08442055 -0.36693838  0.        ] (lightning)
..........coefs: [ 0.11454726 -1.45543426  0.05327869 -0.44623096  0.        ] (LogReg)
L1: 90.00, L2: 10.00
..........coefs: [ 0.         -1.33131389  0.         -0.25033419  0.        ] (SGD)
..........coefs: [ 0.         -1.33131343  0.         -0.25033405  0.        ] (lightning)
..........coefs: [ 0.         -1.38744357  0.         -0.26100491  0.        ] (LogReg)
L1: 100.00, L2: 0.00
..........coefs: [ 0.         -1.3560452   0.         -0.20918346  0.        ] (SGD)
..........coefs: [ 0.         -1.35604451  0.         -0.20918325  0.        ] (lightning)
..........coefs: [ 0.         -1.35604451  0.         -0.20918325  0.        ] (LogReg)
```
![0 0](https://user-images.githubusercontent.com/11065596/47579179-90358000-d94b-11e8-9f1f-0dfce8046845.png)
![0 1](https://user-images.githubusercontent.com/11065596/47579181-90358000-d94b-11e8-899a-e8c9b6f4bce7.png)
![0 5](https://user-images.githubusercontent.com/11065596/47579182-90358000-d94b-11e8-9f6b-6f402c796bb6.png)
![0 9](https://user-images.githubusercontent.com/11065596/47579184-90358000-d94b-11e8-8871-0e1b63ec8faf.png)
![1 0](https://user-images.githubusercontent.com/11065596/47579185-90358000-d94b-11e8-93eb-b42febd1f3f2.png)
___
**With the factor 2:**
```
L1: 0.00, L2: 100.00
..........coefs: [ 0.32072801 -1.14099813  0.24350405 -0.45900174 -0.13345408] (SGD)
..........coefs: [ 0.32072806 -1.14099771  0.24350422 -0.45900164 -0.13345406] (lightning)
..........coefs: [ 0.32072806 -1.14099771  0.24350422 -0.45900164 -0.13345406] (LogReg)
L1: 10.00, L2: 90.00
..........coefs: [ 0.28944688 -1.14915421  0.21572344 -0.44261878 -0.09801055] (SGD)
..........coefs: [ 0.28944694 -1.14915376  0.21572362 -0.44261866 -0.09801052] (lightning)
..........coefs: [ 0.28925892 -1.14712962  0.21570674 -0.44179261 -0.09775064] (LogReg)
L1: 50.00, L2: 50.00
..........coefs: [ 0.14099022 -1.21314473  0.08441194 -0.36694632  0.        ] (SGD)
..........coefs: [ 0.14099785 -1.21312388  0.08442055 -0.36693838  0.        ] (lightning)
..........coefs: [ 0.14107177 -1.21228901  0.08453823 -0.36667399  0.        ] (LogReg)
L1: 90.00, L2: 10.00
..........coefs: [ 0.         -1.33131389  0.         -0.25033419  0.        ] (SGD)
..........coefs: [ 0.         -1.33131343  0.         -0.25033405  0.        ] (lightning)
..........coefs: [ 0.         -1.33128569  0.         -0.25033298  0.        ] (LogReg)
L1: 100.00, L2: 0.00
..........coefs: [ 0.         -1.3560452   0.         -0.20918346  0.        ] (SGD)
..........coefs: [ 0.         -1.35604451  0.         -0.20918325  0.        ] (lightning)
..........coefs: [ 0.         -1.35604451  0.         -0.20918325  0.        ] (LogReg)
```
![0 0b](https://user-images.githubusercontent.com/11065596/47579136-77c56580-d94b-11e8-9d17-64322120f985.png)
![0 1b](https://user-images.githubusercontent.com/11065596/47579137-77c56580-d94b-11e8-948c-78d66de9d49d.png)
![0 5b](https://user-images.githubusercontent.com/11065596/47579138-77c56580-d94b-11e8-8819-fc76be57b806.png)
![0 9b](https://user-images.githubusercontent.com/11065596/47579139-77c56580-d94b-11e8-8864-f9347acae4ca.png)
![1 0b](https://user-images.githubusercontent.com/11065596/47579140-77c56580-d94b-11e8-8f0d-1aa14b73391b.png)

amueller commented:
awesome, thanks for the analysis. Could you also include liblinear? That should also work for l1 and l2 right?

fabianp commented:
Yes, most often SAG is slightly faster in the smooth case.

GaelVaroquaux commented:
> Yes, most often SAG is slightly faster in the smooth case.

Do you have a gut feeling of the ratio?

fabianp commented:
Comparisons of these methods are always tricky since at the end it boils
down to how the step size is selected. For a step size of the form 1/3L,
which is similar to the one used by scikit-learn (and for which SAGA has
guaranteed convergence but not SAG), in my experience and without data at
hand to back this up I would say that the runtime of SAG would be around
0.8 that of SAGA. Would love to hear other's opinion/experience.

On Sun, Oct 28, 2018, 17:57 Gael Varoquaux <notifications@github.com wrote:

> > Yes, most often SAG is slightly faster in the smooth case.
>
> Do you have a gut feeling of the ratio?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/11646#issuecomment-433745206>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAQ8hx2zD9S1KbmUQrgAiFr-1y8i13E3ks5upihJgaJpZM4VZYx1>
> .
>

amueller commented:
fyi I'm not necessarily against dropping SAG I just wanted to confirm.

TomDLT commented:
> awesome, thanks for the analysis. Could you also include liblinear? That should also work for l1 and l2 right?

I included `LogisticRegression(solver='liblinear')` in the benchmark, and it does optimize the same objective function (though only for L1 or L2, not ElasticNet). (see script and figures in the details)

> ok so this suggests a bug in the enet SAGA code right?

Yes. Our SAGA code was not implemented with ElasticNet in mind, but for either L1 or L2. We assumed in this PR that we could call our SAGA implementation with both penalties at the same time, but it does not work. Or more precisely, it does not optimize the desired objective function. Note that this is a problem of our implementation, and not of SAGA, since lightning's SAGA works correctly.

<details>


![0 0](https://user-images.githubusercontent.com/11065596/47641635-dadf1400-db66-11e8-9d99-52776429a255.png)
![1 0](https://user-images.githubusercontent.com/11065596/47641636-db77aa80-db66-11e8-9a6b-11d67c9f38cd.png)


```py
import numpy as np
import matplotlib.pyplot as plt

from sklearn.datasets import make_classification
from sklearn.linear_model import LogisticRegression
from sklearn.linear_model import SGDClassifier
from sklearn.preprocessing import scale
from sklearn.metrics import log_loss
from sklearn.utils.testing import ignore_warnings

from lightning.classification import SAGAClassifier

###############################################################################
# Dataset
n_samples = 1000
X, y = make_classification(n_samples=n_samples, n_classes=2, n_features=5,
                           n_informative=5, n_redundant=0, n_repeated=0,
                           random_state=0)
X = scale(X)

###############################################################################
# Regularization
C = 0.01
for l1_ratio in [0, 0.1, 0.5, 0.9, 1]:

    l2_reg = (1. / C) * (1 - l1_ratio)
    l1_reg = (1. / C) * l1_ratio
    print("L1: %.2f, L2: %.2f" % (l1_reg, l2_reg))

    # parameters for SGD:
    alpha_sgd = (l2_reg + l1_reg) / n_samples
    l1_ratio_sgd = l1_reg / (l2_reg + l1_reg)

    # parameters for lightning's SAGAClassifier
    alpha_lightning = l2_reg / n_samples
    beta_lightning = l1_reg / n_samples

    # parameters for LogisticRegression
    # XXX: weird, somewhat helps getting a better equivalence between SAGAs
    if True:
        l2_reg_ = l2_reg * 2 if l1_reg != 0 else l2_reg
    else:
        l2_reg_ = l2_reg

    C_logistic = 1. / (l2_reg_ + l1_reg)
    l1_ratio_logistic = l1_reg / (l2_reg_ + l1_reg)

    def objective(estimator, X, y):
        coef = estimator.coef_.ravel()
        obj = log_loss(y, estimator.predict_proba(X)) * n_samples
        obj += l1_reg * np.sum(np.abs(coef))
        obj += l2_reg * 0.5 * np.dot(coef, coef)
        return obj

    ###########################################################################
    # Compared methods: (name, max_iter, estimator)
    methods = [
        ('SGD', 10000, SGDClassifier(
            penalty='elasticnet', loss='log', random_state=0, alpha=alpha_sgd,
            l1_ratio=l1_ratio_sgd, tol=-np.inf, fit_intercept=False), ),
        ('SAGA (lightning)', 100,
         SAGAClassifier(alpha=alpha_lightning, beta=beta_lightning, loss='log',
                        tol=-1, penalty='l1', random_state=0), ),
        ('SAGA (sklearn)', 100, LogisticRegression(
            penalty='elasticnet', solver='saga', random_state=0, C=C_logistic,
            l1_ratio=l1_ratio_logistic, tol=0, fit_intercept=False), ),
    ]
    if l1_reg == 0:
        methods.append(('Liblinear', 100, LogisticRegression(
            penalty='l2', solver='liblinear', random_state=0, C=C_logistic,
            l1_ratio=l1_ratio_logistic, tol=1e-15, fit_intercept=False), ))
        methods.append(('SAG (sklearn)', 100, LogisticRegression(
            penalty='l2', solver='sag', random_state=0, C=C_logistic,
            l1_ratio=l1_ratio_logistic, tol=0, fit_intercept=False), ))
    if l2_reg == 0:
        methods.append(('Liblinear', 100, LogisticRegression(
            penalty='l1', solver='liblinear', random_state=0, C=C_logistic,
            l1_ratio=l1_ratio_logistic, tol=1e-15, fit_intercept=False), ))

    ###########################################################################
    # Benchmark: (name, max_iter, estimator)
    results = []
    for name, max_max_iter, estimator in methods:
        max_iter_range = np.int_(np.logspace(0, np.log10(max_max_iter), 10))
        one_result = []
        for max_iter in max_iter_range:
            print('.', end='', flush=True)
            estimator.max_iter = max_iter
            with ignore_warnings():
                estimator.fit(X, y)
            one_result.append(objective(estimator, X, y))

        coef = estimator.coef_.ravel()
        print("coefs: %s (%s)" % (coef, name))
        results.append(np.array(one_result))

    results = np.array(results)
    fig = plt.figure()
    plt.semilogy(max_iter_range, results.T - np.min(results) + 1e-15, '.-')
    plt.legend([name for name, _, _ in methods])
    plt.title('l1_ratio = %s' % l1_ratio)
    plt.xlabel('n_iter (SAGA) or n_iter / 100 (SGD)')

    fig.savefig('%.1f.png' % l1_ratio)
    plt.close('all')
```

TomDLT committed:
FIX incorrect lagged_update in SAGA

TomDLT committed:
Add non-regression test for SAGA's bug

TomDLT committed:
FIX flake8 and warning

NicolasHug committed:
Re fixed warning

NicolasHug commented:
Thanks a lot everyone for your help, especially @TomDLT !

I'll go through the tests and try to make them more thorough now that the bug is fixed.

NicolasHug committed:
Updated some tests

NicolasHug committed:
Merge branch 'master' into elastic_net_linear_classifier

TomDLT reviewed (approved):
Mainly nitpicks.

You will also need a whatsnew entry.
You can also mention the bugfix in saga solver, and add `LogisticRegression(solver='saga')` in the changed models list.

NicolasHug committed:
Addressed comments

NicolasHug committed:
more comments and added dimension to LogisticRegressionCV.n_iter_ attribute

NicolasHug committed:
Updated whatsnew for 0.21

NicolasHug committed:
better doc shape looks

NicolasHug committed:
Merge branch 'elastic_net_linear_classifier' of github.com:NicolasHug/scikit-learn into elastic_net_linear_classifier

NicolasHug committed:
Merge branch 'master' into elastic_net_linear_classifier

NicolasHug committed:
Fixed whatnew entry after merges

NicolasHug commented on doc/whats_new/v0.21.rst:
```diff
@@ -137,6 +139,15 @@ Support for Python 3.4 and below has been officially dropped.
+  with the 'saga' solver. :issue:`11646` by :user:`Nicolas Hug <NicolasHug>`.
+
+- |Fix| Fixed a bug in the 'saga' solver where the weights would not be
+  correctly updated in some cases :issue:`11646` by `Tom Dupre la Tour`_.
```
Thanks for the review @TomDLT , please feel free to modify this entry for a more accurate description. I don't understand the bug fix in details.

jnothman replied:
Add a . after "some cases"

TomDLT reviewed (approved):
Thanks !

amueller commented on doc/modules/linear_model.rst:
```diff
@@ -799,8 +813,8 @@ In a nutshell, the following table summarizes the penalties supported by each so
 
-The "saga" solver is often the best choice but requires scaling. The "liblinear" solver is
-used by default for historical reasons.
+The "saga" solver is often the best choice but requires scaling. The
```
Unrelated but we're moving the default to lbfgs and there is no mention of this in the summary. That's a different PR though.

amueller commented on doc/modules/linear_model.rst:
```diff
@@ -839,13 +853,13 @@ with 'log' loss.
-"multinomial", an optimal C is obtained by minimizing the cross-entropy
-loss.
+multiclass case, if `multi_class` option is set to "ovr", an optimal C and
+l1_ratio is obtained for each class and if the `multi_class` option is set to
```
Wait, we're not using accuracy here or scoring? That seems odd. Is that wording really correct?

amueller replied:
@NicolasHug can you please fix this, even though it wasn't your error?

jnothman reviewed (commented):
A nitpick...

NicolasHug committed:
Added dot

amueller commented on sklearn/linear_model/tests/test_logistic.py:
```diff
@@ -1380,6 +1395,273 @@ def test_warm_start_converge_LR():
+@pytest.mark.parametrize('C', np.logspace(-3, 2, 4))
+@pytest.mark.parametrize('l1_ratio', [.1, .5, .9])
+def test_elastic_net_versus_sgd(C, l1_ratio):
+    # Compare elasticnet penatly in LogisticRegression() and SGD(loss='log')
```
penalty

amueller replied:
not fixed

NicolasHug replied:
Sorry yeah, will do it tomorrow!

amueller commented on sklearn/linear_model/tests/test_logistic.py:
```diff
@@ -1380,6 +1395,273 @@ def test_warm_start_converge_LR():
+                                cv=n_folds, l1_ratios=l1_ratios,
+                                random_state=0)
+    lrcv.fit(X, y)
+    coefs_paths = np.asarray(list(lrcv.coefs_paths_.values()))
```
if l1_ratios is not specified, do we squeeze? Otherwise this would be a backward-incompatible change, right?

amueller replied:
not fixed yet

NicolasHug replied:
We don't need to squeeze because if `l1_ratios` is None then we just don't even add the new dimension (see [here](https://github.com/scikit-learn/scikit-learn/pull/11646/files#diff-1452d97adf1a0b794930e405aa58d64bR2005)).

The existing test that check the shapes of the attributes is left unchanged (`test_logistic_cv`)

amueller commented on sklearn/linear_model/tests/test_logistic.py:
```diff
@@ -1380,6 +1395,273 @@ def test_warm_start_converge_LR():
+    # make sure LogisticRegressionCV gives same best params (l1 and C) as
+    # GridSearchCV when penalty is elasticnet
+
+    if multi_class == 'ovr':
```
so binary?

NicolasHug replied:
yes, multiclass ovr is treated below in `test_LogisticRegressionCV_GridSearchCV_elastic_net_ovr`, just added a comment

amueller commented on sklearn/linear_model/logistic.py:
```diff
@@ -1788,69 +1873,88 @@ def fit(self, X, y, sample_weight=None):
+        # _log_reg_scoring_path will output different shapes depending on the
+        # multi_class param, so we need to reshape the outputs accordingly.
+        # After reshaping,
+        # - scores is of shape (n_classes X n_folds X n_Cs . n_l1_ratios)
```
What's the shape of Cs and why do we always have to take C[0]?

NicolasHug replied:
Cs is of shape `(n_classes . n_folds . n_l1_ratios, n_Cs)` and all the rows are equal, so we just take the first one.

I added a comment.

amueller commented on sklearn/linear_model/logistic.py:
```diff
@@ -1788,69 +1873,88 @@ def fit(self, X, y, sample_weight=None):
+            )
+            # equiv to coefs_paths = np.moveaxis(coefs_paths, (0, 1, 2, 3),
+            #                                                 (1, 2, 0, 3))
+            coefs_paths = np.swapaxes(coefs_paths, 0, 1)
```
you find two swapaxes easier to read than one rollaxis? I don't actually have an opinion as the comment above makes it pretty clear.

NicolasHug replied:
I did at the time when I was trying to break it down but not anymore ^^

amueller reviewed (commented):
looks good apart from nitpicks.

NicolasHug committed:
Addressed comments + standardized optional default param docstrings

NicolasHug committed:
Addessed comments

amueller commented:
test failing...

NicolasHug committed:
use swapaxes instead of unsupported moveaxis (hopefully fixes tests)

NicolasHug commented:
Aaah... `moveaxis` isn't supported in old numpy version, that's actually why I needed the weird `swapaxes`.

Sorry

NicolasHug committed:
Merge branch 'master' into elastic_net_linear_classifier

GaelVaroquaux commented:
Hurray!!

TomDLT commented:
Great ! :confetti_ball:

jnothman commented:
Exciting!
